csharp/src/Ice/InputStream.cs

```diff
@@ -3087,13 +3072,20 @@ private int readInstance(int index, Action<Value> cb)
                     // been read yet in the case of a circular reference to an
                     // enclosing instance.
                     //
-                    int[] table = _current.indirectionTables[n];
+                    int[] table = _current.IndirectionTableList[n];
                     SliceInfo info = _current.slices[n];
                     info.instances = new Value[table != null ? table.Length : 0];
                     for (int j = 0; j < info.instances.Length; ++j)
                     {
-                        var cj = j;
-                        addPatchEntry(table[j], (Value v) => info.instances[cj] = v);
+                        int index = table[j];
+                        if (_unmarshaledMap.TryGetValue(index, out var obj))
```

Comment:
I guess I'm beating a dead horse here given that we're going to get rid of slice preservation but I suspect we could simplify `readInstance` to just return the object instead of a pair and fix the indirection table to store the objects instead of the object IDs.

Reply:
Fixed as suggested: IndirectionTable is now a Value[] instead of an int[].